docs/authoring/yaml-tasks.md

```diff
@@ -0,0 +1,88 @@
+In YAML, there's no affordance for which inputs go together, which are mutually exclusive, and which ones cause others to become required.
+- Choose human-friendly input names.
+The name will be used in the `input` block.
+- Follow the principle of least surprise when choosing `defaultValue`s.
+None of our YAML experiences offer insight into task defaults right now.
+Users have to trust you won't do something silly.
+- Document your inputs, including a YAML snippet showing how to use it.
+We're improving on the product side with things like IntelliSense in the YAML editor, but at the moment, discovering inputs isn't easy.
+
+### Should a task even be a task?
+
+- Tasks which are merely wrappers around a multi-verb command line aren't valuable.
+Example: we would not ship a task called `Git` with an input `Action` whose values are `clone`, `checkout`, and `push`.
+Customers should be advised to simply add a script step which directly called `git clone` or whatever.
+(We've done this before, and we've always regretted it.)
+- Tasks which can automate the setup a developer would perform on their local machine are great.
+For instance, setting up SSH keys is typically a one-time thing on a dev machine, but has to be repeated in each CI run.
+Also, SSH private keys should be stored as secure files, which can only be downloaded to the agent by virtue of being a task input.
+- Tasks offer the ability to inject pre- and post-steps on the job.
+This can occasionally be the only mechanism to accomplish something: for instance, cleaning up after use of a secure resource.
+
+## Advice for upgrading tasks
+
+Obviously, updating an existing task is _much_ trickier than building a new task.
+If, after reviewing this advice, you can't find a clean way to upgrade, you might consider marking your current task deprecated and offering a new, "green field" task that follows the above guidance.
+Consider that a last resort, though, since it's fairly disruptive for your customers.
+
+### Easing life for YAML users
+
+- YAML exposes the raw input name in a way that the classic editor did not.
+If you don't want to make a breaking change to your task, you can add an `aliases` array to the `input` definition.
+It's an array of strings which will be considered equivalent to the input's real name.
+The first one in the list will usually be suggested by IntelliSense.
+Example: the `InstallSSHKey` task took an input called `hostName` but actually expected a line from a `known_hosts` file.
+In the classic editor, this was clear because the label was "Known hosts entry".
+I added an alias, `knownHostsEntry`, so that it's also clear in YAML what's actually expected.
+- The syntax for referring to an extension-installed task is a little verbose.
+It includes the extension publisher and extension ID, plus the task's `name` (NOT `friendlyName`).
+_TODO: check if we support bare task name when it's unambiguous._
```

Comment:
Probably worthwhile to try uploading two public extensions with the same task name to see if it's allowed. I am not sure of what scope we require for uniqueness. E.g.- does a public task name need to be named uniquely across the service(marketplace) or is it scoped within publisher.

Reply:
Good idea